bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+import argparse
+import os
+import re
+import shlex
+import sys
+import textwrap
+
+import yaml
+
+from typing import Any, Dict, Iterator, List, Mapping, Optional, Sequence, Tuple
+
+
+def main():
+    parser = argparse.ArgumentParser(description="""
+    Process a yaml file and generate the desired calibrations
+    """)
+
+    parser.add_argument('dataDir', type=str, help="Root of data repository")
+    parser.add_argument('specFile', type=str, help="Name of file specifying the work")
+    parser.add_argument('--bootstrap', action="store_true", help="Bootstrap the calibrations?")
+    parser.add_argument('--blocks', type=str, nargs="+", help="Blocks to execute")
+    parser.add_argument('--calib', type=str, help="Name of output calibration directory", default="CALIB")
+    parser.add_argument('--calibTypes', type=str, nargs="+", help="Types of calibs to process")
+    parser.add_argument("--copyMode", choices=["move", "copy", "link", "skip"], default="link",
+                        help="How to move files into calibration directory")
+    parser.add_argument('--force', action="store_true", help="Continue in the face of problems", default=False)
+    parser.add_argument("-j", "--processes", type=int, default=1, help="Number of processes to use")
+    parser.add_argument('--rerun', type=str, help="Name of rerun to use for calib processing")
+    parser.add_argument('--scienceSteps', type=str, nargs="+", help="pipeline steps to execute")
+    parser.add_argument('--verbose', action="store_true", help="How chatty should I be?", default=False)
+
+    args = parser.parse_args()
+
+    bootstrap, calibBlocks, scienceBlocks = processYaml(args.specFile)
+    possibleBlocks = unique(list(calibBlocks.keys()) + list(scienceBlocks.keys()))
+    if args.blocks is None:
+        args.blocks = possibleBlocks
+
+    args.dataDir = os.path.abspath(args.dataDir)
+    if not os.path.exists(args.dataDir):
+        print(f"Warning: {args.dataDir} doesn't exist", file=sys.stderr)
+        if not args.force:
+            sys.exit(1)
+
+    unrecognisedBlocks = set(args.blocks) - set(possibleBlocks)
+    if unrecognisedBlocks:
+        print(f"Unrecognised blocks: '{list(unrecognisedBlocks)}'%s" % (" ignoring" if args.force else ""),
+              file=sys.stderr)
+        if args.verbose:
+            print(f"Possible blocks are {list(possibleBlocks)}", file=sys.stderr)
+        if not args.force:
+            sys.exit(1)
```

Comment:
Now that `generateCommands()` function has been split from `main()` , this block may be called from other functions than `main()`. Because the caller may pass invalid arguments to `generateCommands()`, this block still remains.